// dubbo2.js/es7/index.js
"use strict";
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const js_to_java_1 = __importDefault(require("js-to-java"));
exports.java = js_to_java_1.default;
const context_1 = __importDefault(require("./context"));
exports.Context = context_1.default;
const directly_dubbo_1 = __importDefault(require("./directly-dubbo"));
exports.DirectlyDubbo = directly_dubbo_1.default;
const dubbo_1 = __importDefault(require("./dubbo"));
exports.Dubbo = dubbo_1.default;
const go_1 = require("./go");
exports.go = go_1.go;
const setting_1 = __importDefault(require("./setting"));
exports.setting = setting_1.default;
